src/analyses/libraryFunctions.ml

```diff
@@ -30,18 +35,37 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("pthread_exit", special [__ "retval" []] @@ fun retval -> ThreadExit { ret_val = retval }); (* Doesn't dereference the void* itself, but just passes to pthread_join. *)
+    ("pthread_cond_wait", special [__ "cond" []; __ "lock" []] @@ fun cond lock -> CondWait { cond; lock });
+    ("pthread_cond_timedwait", special [__ "cond" []; __ "lock" []; drop "abstime" [r]] @@ fun cond lock -> CondWait { cond; lock });
   ]
 
 (** GCC builtin functions.
     These are not builtin versions of functions from other lists. *)
 let gcc_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("__builtin_object_size", unknown [drop "ptr" [r]; drop' []]);
+    ("__builtin_prefetch", unknown (drop "addr" [] :: VarArgs (drop' [])));
+    ("__builtin_expect", special [__ "exp" []; drop' []] @@ fun exp -> Identity exp);
+    ("__builtin_unreachable", special' [] @@ fun () -> if get_bool "sem.builtin_unreachable.dead_code" then Abort else Unknown); (* https://github.com/sosy-lab/sv-benchmarks/issues/1296 *)
+    ("__assert_rtn", special [drop "func" [r]; drop "file" [r]; drop "line" []; drop "exp" [r]] @@ Abort); (* gcc's built-in assert *)
+    ("__builtin_return_address", unknown [drop "level" []]);
   ]
 
-(** Linux kernel functions. *)
-let linux_descs_list: (string * LibraryDesc.t) list = (* LibraryDsl. *) [
+let big_kernel_lock = AddrOf (Cil.var (Goblintutil.create_var (makeGlobalVar "[big kernel lock]" intType)))
+let console_sem = AddrOf (Cil.var (Goblintutil.create_var (makeGlobalVar "[console semaphore]" intType)))
 
-]
+(** Linux kernel functions. *)
+(* TODO: conditional on kernel option *)
+let linux_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
+    ("spin_lock_irqsave", special [__ "lock" []; drop "flags" []] @@ fun lock -> Lock { lock; try_ = get_bool "sem.lock.fail"; write = true; return_on_success = true });
+    ("spin_unlock_irqrestore", special [__ "lock" []; drop "flags" []] @@ fun lock -> Unlock lock);
+    ("_raw_spin_unlock_irqrestore", special [__ "lock" []; drop "flags" []] @@ fun lock -> Unlock lock);
+    ("spinlock_check", special [__ "lock" []] @@ fun lock -> Identity lock);
+    ("_lock_kernel", special [drop "func" [r]; drop "file" [r]; drop "line" []] @@ Lock { lock = big_kernel_lock; try_ = false; write = true; return_on_success = true });
+    ("_unlock_kernel", special [drop "func" [r]; drop "file" [r]; drop "line" []] @@ Unlock big_kernel_lock);
+    ("acquire_console_sem", special [] @@ Lock { lock = console_sem; try_ = false; write = true; return_on_success = true });
+    ("release_console_sem", special [] @@ Unlock console_sem);
+    ("misc_deregister", unknown [drop "misc" [r_deep; s_deep]]); (* TODO: why does this need to spawn to pass 04-mutex/40-rw_lock_rc? *)
```

Comment:
Some races in that test arise from `misc_deregister(&test_dev);` calling functions from the device being freed, but there's also `misc_register(&test_dev);` which should already be doing that as well, so why is this necessary?

Reply:
Yes, de-register should not need to spawn these, but if only register spawns, we get the following accesses on `data1`:

```
[Success][Race] Memory location data1@tests/regression/04-mutex/40-rw_lock_rc.c:8:12-8:17 (safe):
  write with [lock:{read lock rwlock1}, thread:[test_init, my_read@tests/regression/04-mutex/40-rw_lock_rc.c:51:3-51:33]] (conf. 110) (tests/regression/04-mutex/40-rw_lock_rc.c:20:3-20:21)
  write with [lock:{rwlock1}, thread:[test_init, my_write@tests/regression/04-mutex/40-rw_lock_rc.c:51:3-51:33]] (conf. 110) (tests/regression/04-mutex/40-rw_lock_rc.c:27:3-27:27)
```

This is without MHP, but it's similar with it... This test was created before we had this pairwise view and was based on the heuristics that you shouldn't write with a read lock, but this is not a problem in itself. In this case, the above accesses are not conflicting if the threads are unique, but this access on line 20 should conflict with itself or do we asssume unknown-function spawns are unique?

Reply:
I made it non-spawning and changed the test to use plain (not history) thread IDs instead, so that it doesn't have the uniqueness information. Unknown function spawns are a good point, tracked now by #843.